### PR TITLE
Refactor rate model

### PR DESCRIPTION
**Pull request by yz89**
- Refactor rate model #32 #60
- Add CHANGELOG.md #75

### CODE REVIEW

File: pallets/loans/src/lib.rs [hunk 1]
```diff
@@ -554,4 +563,47 @@ impl<T: Config> Pallet<T> {
 
         Ok(())
     }
+
+    pub fn calc_utilization_ratio(
+        cash: Balance,
+        borrows: Balance,
+        reserves: Balance,
+    ) -> Result<Ratio, Error<T>> {
+        // utilization rate is 0 when there are no borrows
```
Comment (GopherJ) [inline pallets/loans/src/lib.rs:572]: should be ratio

File: pallets/loans/src/lib.rs [hunk 2]
```diff
@@ -554,4 +563,47 @@ impl<T: Config> Pallet<T> {
...
+        if borrows.is_zero() {
+            return Ok(Ratio::zero());
+        }
+        // utilizationRate = totalBorrows / (totalCash + totalBorrows − totalReserves)
```
Comment (GopherJ) [inline pallets/loans/src/lib.rs:576]: should be ratio

File: pallets/loans/src/rate.rs [hunk 3]
```diff
@@ -12,105 +12,162 @@
...
-
-        let multiplier_per_block = MultiplierPerBlock::<T>::get();
-        let base_rate_per_block = BaseRatePerBlock::<T>::get();
-        let jump_multiplier_per_block = JumpMultiplierPerBlock::<T>::get();
-        let kink = Kink::<T>::get();
-
-        if util <= kink {
-            let rate = multiplier_per_block
+    /// Calculates the current borrow interest rate per block
+    pub fn get_borrow_rate(&self, util: Ratio) -> Option<Rate> {
```
Comment (GopherJ) [inline pallets/loans/src/rate.rs:57]: can we have a formula here?

**Review by GopherJ — APPROVED** [review]

**Comment by GopherJ** [comment]
The code looks much more clear now! Thanks @yz89.  Another advice is to have formula in every important place:

- utilization ration
- borrow rate
- supply rate
- accure interest
- ...

the will help new team member

File: pallets/loans/src/lib.rs [hunk 4]
```diff
@@ -538,6 +523,30 @@ impl<T: Config> Pallet<T> {
         T::PalletId::get().into_account()
     }
 
+    fn accrue_interest() -> DispatchResult {
+        for currency_id in Self::currencies() {
+            let total_cash = Self::get_total_cash(currency_id);
+            let total_borrows = Self::total_borrows(currency_id);
+            let util = Self::calc_utilization_ratio(total_cash, total_borrows, 0)?;
```
Comment (ayushmishra2005) [inline pallets/loans/src/lib.rs:530]: Can we replace `0` with `Zero::zero()` everywhere?

**Comment by yz89** [comment]
> The code looks much more clear now! Thanks @yz89. Another advice is to have formula in every important place:
> 
> * utilization ration
> * borrow rate
> * supply rate
> * accure interest
> * ...
> 
> the will help new team member

Thanks, I will improve it.

File: pallets/loans/src/lib.rs (hunk 4 above)
Comment (yz89) [reply pallets/loans/src/lib.rs:530]: Feasible. Good advice.

**Review by ayushmishra2005 — APPROVED** [review]

**Pull request merged**